File: sdk/kotlin/src/test/kotlin/com/loomup/client/OfflineStoreTest.kt
package com.loomup.client

import kotlinx.coroutines.runBlocking
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import java.io.File
import kotlin.io.path.createTempFile
import kotlin.test.Test
import kotlin.test.assertEquals

@Serializable
private data class OfflineFixture(
    val bootstrap: SyncBootstrapResponse,
    @SerialName("offline_mutations") val offlineMutations: List<SyncMutation>,
    @SerialName("mutation_response") val mutationResponse: SyncMutationResponse,
    val pull: SyncPullResponse,
    val expected: Expected,
) {
    @Serializable data class Expected(val cursor: Long, val pending: Int, val conflicts: Int, val ids: List<String>)
}

private class FixtureSyncTransport(private val fixture: OfflineFixture) : SyncTransport {
    override suspend fun syncBootstrap(resources: List<String>, clientId: String) = fixture.bootstrap
    override suspend fun syncPull(cursor: Long, resources: List<String>, clientId: String) = fixture.pull
    override suspend fun syncMutations(mutations: List<SyncMutation>) = SyncMutationResponse(
        1, fixture.mutationResponse.results.filter { it.mutationId == mutations.single().id },
    )
}

class OfflineStoreTest {
    @Test
    fun sqliteStoragePersistsState() = runBlocking {
        val file = createTempFile("loomup-", ".sqlite").toFile()
        try {
            SQLiteSyncStorage(file).use { storage ->
                storage.setItem("state", "durable")
                assertEquals("durable", storage.getItem("state"))
                storage.removeItem("state")
                assertEquals(null, storage.getItem("state"))
            }
        } finally { file.delete() }
    }

    @Test
    fun sharedOfflineV1QueueReconnectConformance() = runBlocking {
        val fixture = Json { ignoreUnknownKeys = true }.decodeFromString<OfflineFixture>(
            File("../conformance/offline-v1.json").readText(),
        )
        val store = OfflineStore.open(FixtureSyncTransport(fixture), listOf("items"))

        store.setOnline(false)
        val create = fixture.offlineMutations[0]
        store.create(create.resource, create.data!!, create.recordId, create.id)
        val update = fixture.offlineMutations[1]
        store.update(update.resource, update.recordId!!, update.data!!, update.id)
        assertEquals(2, store.status.value.pending)
        assertEquals(OfflinePhase.OFFLINE, store.status.value.phase)

        store.setOnline(true)
        assertEquals(fixture.expected.cursor, store.status.value.cursor)
        assertEquals(fixture.expected.pending, store.status.value.pending)
        assertEquals(fixture.expected.conflicts, store.status.value.conflicts)
        assertEquals(fixture.expected.ids.sorted(), store.find("items").mapNotNull { it["id"]?.stringValue }.sorted())
    }
}

File: sdk/flutter/test/offline_store_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:loomup/loomup.dart';
import 'package:test/test.dart';

class FixtureTransport implements SyncTransport {
  final Map<String, dynamic> fixture;
  FixtureTransport(this.fixture);

  @override
  Future<SyncBootstrapResponse> syncBootstrap(
          List<String> resources, String clientId) async =>
      SyncBootstrapResponse.fromJson(
          Map<String, dynamic>.from(fixture['bootstrap'] as Map));

  @override
  Future<SyncPullResponse> syncPull(
          int cursor, List<String> resources, String clientId) async =>
      SyncPullResponse.fromJson(
          Map<String, dynamic>.from(fixture['pull'] as Map));

  @override
  Future<SyncMutationResponse> syncMutations(
      List<SyncMutation> mutations) async {
    final results = (fixture['mutation_response']['results'] as List)
        .where((value) => (value as Map)['mutation_id'] == mutations.single.id)
        .map((value) => SyncMutationResult.fromJson(
            Map<String, dynamic>.from(value as Map)))
        .toList();
    return SyncMutationResponse(1, results);
  }
}

class FakeSQLiteDatabase implements SQLiteSyncDatabase {
  final Map<String, String> values = {};
  @override
  Future<void> execute(String sql,
      [List<Object?> parameters = const []]) async {
    if (sql.startsWith('INSERT')) {
      values[parameters[0] as String] = parameters[1] as String;
    }
    if (sql.startsWith('DELETE')) {
      values.remove(parameters[0]);
    }
  }

  @override
  Future<List<Map<String, Object?>>> query(String sql,
      [List<Object?> parameters = const []]) async {
    final value = values[parameters[0]];
    return value == null
        ? []
        : [
            {'value': value}
          ];
  }
}

void main() {
  test('SQLite adapter persists state behind the structural API', () async {
    final storage = await SQLiteSyncStorage.open(FakeSQLiteDatabase());
    await storage.setItem('state', 'durable');
    expect(await storage.getItem('state'), 'durable');
    await storage.removeItem('state');
    expect(await storage.getItem('state'), isNull);
  });

  test('shared offline v1 queue reconnect conformance', () async {
    final fixture = Map<String, dynamic>.from(
        jsonDecode(File('../conformance/offline-v1.json').readAsStringSync())
            as Map);
    final mutations = fixture['offline_mutations'] as List;
    final store = await OfflineStore.open(
        transport: FixtureTransport(fixture), resources: ['items']);

    await store.setOnline(false);
    final create =
        SyncMutation.fromJson(Map<String, dynamic>.from(mutations[0] as Map));
    await store.create(create.resource, create.data!,
        recordId: create.recordId, mutationId: create.id);
    final update =
        SyncMutation.fromJson(Map<String, dynamic>.from(mutations[1] as Map));
    await store.update(update.resource, update.recordId!, update.data!,
        mutationId: update.id);

    expect(store.status.pending, 2);
    expect(store.status.phase, OfflinePhase.offline);
    await store.setOnline(true);

    final expected = Map<String, dynamic>.from(fixture['expected'] as Map);
    expect(store.status.cursor, expected['cursor']);
    expect(store.status.pending, expected['pending']);
    expect(store.status.conflicts, expected['conflicts']);
    expect(store.find('items').map((row) => row['id']).toList()..sort(),
        List<String>.from(expected['ids'] as List)..sort());
    await store.close();
  });
}
